vncviewer/touch.cxx

```diff
@@ -0,0 +1,627 @@
+  prepXEvent(&fakeEvent, origEvent);
+
+  switch (button) {
+    case Button1:
+      mask = Button1Mask;
+      break;
+    case Button2:
+      mask = Button2Mask;
+      break;
+    case Button3:
+      mask = Button3Mask;
+      break;
+    case Button4:
+      mask = Button4Mask;
+      break;
+    case Button5:
+      mask = Button5Mask;
+      break;
+    case Button6:
+      mask = Button6Mask;
+      break;
+    case Button7:
+      mask = Button7Mask;
+      break;
+    default:
+      mask = 0;
+  }
+
+  if (press)
+    fakeEvent.xbutton.state |= mask;
+  else
+    fakeEvent.xbutton.state &= ~mask;
+
+  fl_handle(fakeEvent);
+}
+
+void FLTKGestureHandler::fakeKeyEvent(bool press, int keysym,
+                                      const GHEvent origEvent)
+{
+  XEvent fakeEvent;
+
+  Window root, child;
+  int x_root, y_root;
+  XkbStateRec state;
+
+  root = XDefaultRootWindow(fl_display);
+  XTranslateCoordinates(fl_display, wnd, root,
+                        origEvent.event_x,
+                        origEvent.event_y,
+                        &x_root, &y_root, &child);
+  XkbGetState(fl_display, XkbUseCoreKbd, &state);
+
+  KeyCode kc = XKeysymToKeycode(fl_display, keysym);
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = press ? KeyPress : KeyRelease;
+  fakeEvent.xkey.type = press ? KeyPress : KeyRelease;
+  fakeEvent.xkey.keycode = kc;
+  fakeEvent.xkey.serial = XLastKnownRequestProcessed(fl_display);
+  fakeEvent.xkey.display = fl_display;
+  fakeEvent.xkey.window = wnd;
+  fakeEvent.xkey.root = root;
+  fakeEvent.xkey.subwindow = None;
+  fakeEvent.xkey.time = fl_event_time;
+  fakeEvent.xkey.x = origEvent.event_x;
+  fakeEvent.xkey.y = origEvent.event_y;
+  fakeEvent.xkey.x_root = x_root;
+  fakeEvent.xkey.y_root = y_root;
+  //fakeEvent.xkey.state = state.mods;
+  //fakeEvent.xkey.state |= ((state.ptr_buttons >> 1) & 0x1f) << 8; // FIXME: Check this!
+  fakeEvent.xkey.same_screen = True; // FIXME
+
+  XSendEvent(fl_display, wnd, true, 0, &fakeEvent);
+}
+
+void FLTKGestureHandler::handleGestureEvent(const GHEvent& ev)
+{
+  switch (ev.type) {
+  case GH_GestureBegin:
+    switch (ev.gesture) {
+    case GH_LEFTBTN:
+      vlog.info("Got GH_GestureBegin(GH_LEFTBTN)");
+      fakeButtonEvent(true, Button1, ev);
+      break;
+    case GH_MIDDLEBTN:
+      vlog.info("Got GH_GestureBegin(GH_MIDDLEBTN)");
+      fakeButtonEvent(true, Button2, ev);
+      break;
+    case GH_RIGHTBTN:
+      vlog.info("Got GH_GestureBegin(GH_RIGHTBTN)");
+      fakeButtonEvent(true, Button3, ev);
+      break;
+    case GH_VSCROLL:
+      vlog.info("Got GH_GestureBegin(GH_VSCROLL)");
+      lastDistance = ev.detail;
+      break;
+    case GH_HSCROLL:
+      vlog.info("Got GH_GestureBegin(GH_HSCROLL)");
+      lastDistance = ev.detail;
+      break;
+    case GH_ZOOM:
+      vlog.info("Got GH_GestureBegin(GH_ZOOM)");
+      lastDistance = ev.detail;
+      fakeKeyEvent(true, XK_Control_L, ev);
```

Comment:
Have you tried moving this so that it fakes Ctrl on each scroll instead? It would avoid all the issues of having Ctrl constantly pressed down, which could be very difficult to fake completely.

Reply:
That would be very inefficient, though. Two fake keypresses per scrollwheel event, with all the XKB and X11 calls we need to fake them...

Reply:
Fixed